presentation/frontend/src/presentation-frontend/PresentationManager.ts

```diff
@@ -192,10 +192,24 @@ export class PresentationManager implements IDisposable {
       return [];
 
     const options = await this.addRulesetAndVariablesToOptions(props);
-    let modifications: PartialHierarchyModification[];
+    let modifications: PartialHierarchyModification[] = [];
+
     try {
-      modifications = (await this.rpcRequestsHandler.compareHierarchies(this.toRpcTokenOptions(options)))
-        .map(PartialHierarchyModification.fromJSON);
+      while (true) {
+        const result = (await this.rpcRequestsHandler.compareHierarchies(this.toRpcTokenOptions(options)));
+        modifications.push(...result.changes.map(PartialHierarchyModification.fromJSON));
+        if (!result.nextStep)
+          break;
+
+        const prevStep = options.startPosition;
+        const nextStep = result.nextStep;
+        if (prevStep !== undefined && prevStep.prevHierarchyNode === nextStep.prevHierarchyNode && prevStep.currHierarchyNode === nextStep.currHierarchyNode) {
+          Logger.logError(PresentationFrontendLoggerCategory.Package, "Hierarchy compare next step is the same as previous. Returning [].");

```

Comment:
Do you think this check really helps? It doesn't look like we can handle all the cases when backend returns crappy results. E.g. now you handle when 2 consecutive requests return the same `startPosition`, but the check wouldn't work if you get A -> B -> A -> B, etc.

Reply:
I wanted to have atleast some ability to catch incorrect results from backend and avoid infinite loops. Changed this check to log error and break loop if backend returns no changes but valid continuation token.